Word Ladder/solution.cpp

```diff
@@ -0,0 +1,74 @@
+#include "solution.h"
+
+int LadderLength(std::string begin_word, std::string end_word,
+                 std::vector<std::string>& word_list) {
+  // Find shortest sequence to reach end_word -> Graph problem
+  // Shortest path is guaranteed by BFS. DFS would also find a path, but does
+  // not guarantee the shortest path. Time: O(N * M^2) Space: O(N * M^2), where
+  // N - words in word_list, M - length of words in word_list M^2 comes from
+  // finding all possible 1 letter changes in a word, and then we need to store
+  // them in neighbors_map (hash map)
+
+  // Base case: end_word is NOT in word_list
+  if (find(word_list.begin(), word_list.end(), end_word) == word_list.end()) {
+    return 0;
+  }
+
+  // Create hashMap to store possible paths for every letter change
+  // for each word
+  absl::flat_hash_map<std::string, std::vector<std::string> > neighbors_map;
+  int word_len = begin_word.length();  // words are of equal length
+
+  // Go through each word and create store possible patterns
+  // associated with a given word
+  // E.g neighbors_map[h*t] = {hot, hit}
+  for (std::string word_ : word_list) {
+    for (int i = 0; i < word_len; ++i) {
+      std::string pattern_ =
+          word_.substr(0, i) + '*' + word_.substr(i + 1, word_len);
+      neighbors_map[pattern_].push_back(word_);
+    }
+  }
+
+  // Run BFS on neighbors_map. If we reached the word, return shortest_path
+  // If can't, while-loop will eventually run out of element in words_queue 
+  // and return 0
+  int shortest_path = 1;
+  std::set<std::string> visited_set;
```

Comment:
you can also use flat_hash_set.
It's faster and has .contains()

Reply:
That's a good point about .contains().
I initially tried it with .find() and it showed 3 times slow-down over std::set, so I switched back to std::set. Now I tested flat_hash_set with .contains() and it works better